### ISSUE TITLE
API compatibility, users, and new development. Goals and future revision

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#237]
Heya :cowboy_hat_face: , 

I want to discuss with you what we want from this repo, and how do we get there. I have been avoiding this decision for a while now, and we've been hitting decision problems with it repeatedly ( #216. #169 , etc). 

## API 

We have the following APIs (and "APIs"): 

### algorithms
 (SP, TM, Encoders,...) used for core HTM functionality (a single Region) 
Have no API per se, but documentation and public methods published at Numenta's docs. 
Problem with the API is that it defines even implementation details, making other development more difficult. 

### Network API 
(Regions, Engine classes) used for constructing a HTM network made of Regions. 
Better defined, quite well designed. 

### Language bindings (Py) 

Language bindings for PY and other languages. 

### OPF 
Used for swarming, constructing network from JSON text files. 

## Users 

I think we should propagate our fork much more, I'd say we already supersede nupic.core (+native Windows support, + moden c++11/17, + speed optimizations, + python 3 support, - API compatibility?) 

### Nupic.cpp

our repo has ~10 viewers and 3 active developers. 

### Numenta's nupic.core 

80-250 viewers, currently not actively developed, in mainteinence only! 

### Numenta's nupic (python) repo 

The largest community, cca 600-6000 viewers, also only in mainteinence now. Does use the c++ backend for speed. 

## API compatibility 

With the docs and Numenta's repos. Advantage of compatibility is that more users can more easily use our code, disadvantage are limitations to development (in cases of wrongly defined APIs, eg "Algorithms") 

Possible ways to go: 

### Strict

Strictly conforms to the APIs, makes possible drop-in replacements, it is more difficult (or impossible) to make more radical changes. 

### Loose 

Easier to develop own ideas, danger of losing users and becoming useless. 

### Moderate-HTM

Proposal, that if we want to take over the active development of nupic.core, we can also define the APIs ourselves. With minimal changes, but do change the API in case it is not well designed or nonexistant.

**Comment by breznak** [comment]
I would like to start pitching our repo more on the nupic forums, discourse, so we can get more relevant feedback from active users. 

Mi liking would be: 
- moderate API compat: ability to hide/change implementation details, but try to make minimal breaking changes and document them. Still move the codebase forward. Allow better fun/work ratio. 
- users, I would be fine even with c++ only code, if enough users are interested, and keeping py bindings for NetworkAPI and core algorithms.

**Comment by dkeeney** [comment]
@breznak  I am totally with you.
I just wonder if we are ready for prime-time yet.  
- Many of the .py unit tests are still not running.
- pickle serialization interface not working yet. (probably breaks most of the above tests)
- I need to finish folding SDR into the NetworkAPI.
- Not a lot of experience at running Python3 on Window against this code base.
- Our unit tests are thin. Could use more end-to-end tests.
- We could use some more how-to-use documentation.
- We still need to port nupic.py to Python3 (preferably so it runs with both Python2.7 and Python3).

We don't want the early adopters to have too bad of an experience but on the other hand, we could keep nit-picking at this code for years.  Perhaps we can announce it as a beta and ask for help on the last four items in the list above.  

I like @ctrl-z-9000-times idea of the survey.  That can be part of the announcement. All or any input from the forum would really help in determining where we go from here.

I do not know what grid cells and the whole TBT thing will mean to us.  Perhaps it will require something so radically different from what we have that we cannot keep anything. Maybe too early too tell.   But once we get the buy-in for our effort from the community than we can remove all constraints of keeping the old Nupic API.

**Comment by dkeeney** [comment]
You say we are ready for C++ only users...but keep in mind that a C++ user that wants to write a app using the NetworkAPI will find that several major parts are missing; most importantly RegionSP and RegionTM.  But for those coding directly from the algorithms we have everything.

How about we finish the skipped .py unit tests and announce that we are open for business.

**Comment by breznak** [comment]
>  a C++ user that wants to write a app using the NetworkAPI will find that several major parts are missing; most importantly RegionSP and RegionTM 

true that. Wel,, for sb coming from numenta cpp we have an improvement. 

> How about we finish the skipped .py unit tests and announce that we are open for business. 

do you suggest holding off the survey? or just marking some beta-ready status?

**Comment by dkeeney** [comment]
Well, I am not sure the bindings work.  I am getting compile errors .. well warnings really ... in py_sparseMatrix and I cannot figure them out.  I don't think it would take much time to get these cleaned up and I think we should at least to that I would think.  This is on the MSVC build; have not checked ubuntu.

**Comment by ctrl-z-9000-times** [comment]
> I just wonder if we are ready for prime-time yet.

TBH I don't think we are either.  Our announcements should make this point clear.  I would still like to post the survey at some point though.

**Comment by ctrl-z-9000-times** [comment]
Here are my arguments for the Loosest API-Compatability:

The algorithms are hard to use, and IMO this is by bad design.  I think if we put a bit thought into how these things fit together that we make algorithms which are a lot easier to use.

For example the C++ SpatialPooler has 17 arguments, don't get them in the wrong order...
* `localAreaDensity` & `numActiveColumnsPerInhArea` do the same thing, and localAreaDensity is more useful because it is independent of the size of the area.  
* `globalInhibition=false` should probably never be used.  It's 70x slower and the unit-tests are missing.  I have a good replacement but it has a different API.
* We could refactor topology into a function and provide a few standard ones, see issue #92. Also `wrapAround` was clearly appended to the end...
* We could refactor boosting somehow.  IMO we should replace it with the logarithmic method of boosting which has one few parameters and just works.
* I doubt that `minPctOverlapDutyCycles` is useful.  I think that boosting and `raisePermanencesToThreshold` should combine to do the same thing that this does.  This needs some testing though.

Frameworks: Currently NetworkAPI shields the users from the complexity of the algorithms.  If we clean up the algorithms then the frameworks will break, but the frameworks also become less important. 

Fixing any of this would break all of the users applications, but at least they could rewrite it more easily.

**Comment by breznak** [comment]
> Here are my arguments for the Loosest API-Compatability:
The algorithms are hard to use, and IMO this is by bad design. I think if we put a bit thought into how these things fit together that we make algorithms which are a lot easier to use.

I agree with this level of compatibility, it is what I consider moderate changes. (don't exactly agree with all the examples you've listed, but that was just an illustration), 
And as you say, NetworkAPI is rather ok...because it's an actual API, the Algorithms are not, it's just a doc of the implementation. 

I'd say it is not such a hard change if some cunstructor parameters change, or params of the "main" methods (compute, initialize, ...). I also think most of the people construct their HTM apps in the networkAPI model (with regions), so they would not be affected by the low level changes. And the low level changes, on the other hand, give the most flexibility to us.

**Comment by ctrl-z-9000-times** [comment]
> We have the following APIs (and "APIs"):

I think you forgot the Python API's...
The python code depends on the C++/Python bindings.  Most of the users depend on Python.

I think the python repo should be part of our project, regardless of what we do with the repo.  Correct me if I'm wrong: Dkeeney thinks we should maintain compatibility with Numenta's python code, which means we  can't really change the python, which is a big constraint on how we can change the API.

**Comment by ctrl-z-9000-times** [comment]
I posted the survey at: https://discourse.numenta.org/t/survey-features-api-compatibility/5344

**Comment by ctrl-z-9000-times** [comment]
Update: We've already gotten responses to the survey, which is great because it shows interest from our potential users.

I would like to leave the survey open for at least 1 week from initial posting before we start drawing conclusions from it, to allow everyone time to respond.  I'm assuming most people check the forum at least once a week.
* Is one week too short? would 2 weeks be better? I feel that 1 month is too long to wait.

A few days before we start seriously looking at the survey results, we should post a notice on the forum:
*  Say something like: "We will draw our initial conclusions from this survey and begin planning based on them, no sooner than DATE TBD.  This survey will remain open for latecomers who wish to be heard."
* Calendar: maybe post this notice on Feb. 5th, and take results from survey sometime between Feb 8-11th?

After we decide what to do about the things in the survey, we should update the forum with our tentative plans.

**Comment by dkeeney** [comment]
Sounds like a good plan. I think 1 week should be long enough.  There might be more entries later than that but probably not enough to affect the results.  So far I like the results.  One surprise is the interest in combining python and C++ into one repository. 

I think Matt's contribution should be weighted larger that most because he probably has a better feel of the community than anyone.

**Comment by dkeeney** [comment]
Yes, we actually have two entirely separate projects already... one for the dependencies (external) and one for our code. Both have separate cache so once built, the external target is not re-built unless something inside external is changed or it is cleared.

The catch is that we need one top level CMakeList.txt file which determines what to do when the repository as a whole is built. 
- for a python only user, we don't do anything except maybe run some .py unit tests.
- for a C++ only user we build the core library, and we run the c++ unit tests.
- for a python/C++ user, we build the NetworkAPI and the .py interfaces, and do both sets of tests.

The action depends on the argument 'BINDING_BUILD' (values none, Python2, Python3, ...)

And that is basically what we have now. So all we need to do is copy in the nupic.py repository and add documentation.

**Comment by breznak** [comment]
> I posted the survey at: https://discourse.numenta.org/t/survey-features-api-compatibility/5344 

Thanks for setting up the poll! :100: 

>  I think 1 week should be long enough.  

Agree. 1 week is OK, and announce the "end" poll on 5th and "close" on 8/Feb.  

> Yes, we actually have two entirely separate projects already... one for the dependencies (external) and one for our code. 

Think David meant project-as a repo, while David means (CMake) project :yum:

**Comment by dkeeney** [comment]
> Think David meant project-as a repo, while David means (CMake) project 

Sorry; I got a little carried away there.... :)   My point is that we can merge the two repos keeping CMake's separate without  a lot of work.

**Comment by Thanh-Binh** [comment]
I did

**Comment by ctrl-z-9000-times** [comment]
Here are the initial results of the survey:

**API-Compatibility:** Moderates won.  For reference, the text is as follows:
> Minor API changes are okay, as long as they're done for good reason, are well documented, and cause minimal disruption.

**Code Organization:** All one repository won.  For reference, the text is as follows:
> Ease of use, better CI, will maintain strict separation between C++ python and bindings.

**Desired Features:**  Here is the current ranking:
* 69% C++ with Python bindings
* 56% Algorithms in C++ 
* 52% Network API
* 43% Python Algorithms
* 30% OPF & Swarming
* 17% Other: CUDA, Python Introspection/Ease of use

**Comment by ctrl-z-9000-times** [comment]
Our VIP rhyolight says: "I just want python bindings to the Network API"

**Comment by breznak** [comment]
yep, what do we draw from it ? :thinking: 

> API-Compatibility: Moderates won. For reference, the text is as follows:

Moderate:46% - Loose:37% - Strict:17% 

So if we take strict=0.0 deviation, loose=1.0 , so we should be like `0.7` changes to satisfy most people. 

> Code Organization: All one repository won.  

This is a clear win! 
(I'll wonder how we'd approach other languages /c#, cuda/ but there are none competitive bindings than py, as of now) 

> Desired Features: Here is the current ranking: 

this is harder to interpret, as some choices are against-each other, or incomplete... 
I'd say we can 
- scratch off: `OPF + other as ununteresting`
- I'd say `Algorithms in Py (no C++) for experimentation` should be left to numenta/nupic / community/nupic.py (so we would not care here in cpp&bindings repo) 
- and we are left with: 
  - focus on fast c++ core
  - with c++ NetworkAPI available
  - and python bindings to that

And I think that exactly ends up as 

> rhyolight says: "I just want python bindings to the Network API"

**Comment by ctrl-z-9000-times** [comment]
> * I'd say `Algorithms in Py (no C++) for experimentation` should be left to numenta/nupic / community/nupic.py (so we would not care here in cpp&bindings repo)

I'd like to keep all of the algorithms which python has and C++ does not.  There are a bunch of python-only encoders which are critical for making an HTM.  I agree that **some** of the Py algorithms should be left behind.

> and we are left with c++ NetworkAPI available

The VIP specifies python not C++ for the NetworkAPI.  The NetworkAPI could use the python bindings and be implemented in python.

**Comment by breznak** [comment]
> keep all of the algorithms which python has and C++ does not. 

oh, definitely, not removing existing useful code. I was thinking about it the other way around, we provide bindings to c++ functionality, and it's up to Py maintainers to do. But if we wanna be both... 

- what with the duplicates (py,cpp)? //would be nice to keep 
- what if the py duplicate hinders cpp devel - remove? 

> specifies python not C++ for the NetworkAPI.

my bad. sorry. 
So python wrapper to cpp networkAPI. 

- do we keep (existing) bindings to lower-level algorithms? (algorithms/SP,TM..) 
- do you want to proceed with your idea of Connections-Py bindings? 

If we focused only on NetworkAPI compatibility, it would make our live a lot easier. I agree for some research ideas you want to reach lower, but then you probably could prototype in Py only, or write it in cpp direcly..?

**Comment by ctrl-z-9000-times** [comment]
> Py maintainers

It's just us chickens.

>    * do we keep (existing) bindings to lower-level algorithms? (algorithms/SP,TM..)
>    * do you want to proceed with your idea of Connections-Py bindings?

It's up to you and dkeeney, since I avoid working on the NetworkAPI.  I'm just pointing out an alternative solution.

**Comment by ctrl-z-9000-times** [comment]
I've created a new milestone "Release version 1.0" to track our progress.  We should use this thread to discuss which issues should be attached to the release v1 milestone.

**Comment by breznak** [comment]
Nice, 

>  attached to the release v1 milestone.

I'd say
- [x] Windows MSVC support
- [ ] C++ NetworkAPI 
- [x] pybind11 bindings
- [x] py2, py3 bindings support
- [x] SP on COnnections
- [ ] Hotgym bentchmark for testing & comparison (+MNIST) 
- [ ] cleandup,removed SparseMatrix (if we agree on that one) 
- [ ] (maybe) finalize Python skipped tests (pickle mostly) 

those are most important to me (relative to their completition status, etc)

**Comment by ctrl-z-9000-times** [comment]
>  SP on COnnections

Do you mean the python SP? or the C++ SP?

> Python skipped tests

There are a lot of python tests which are skipped because they were never copied from the python repo into the C++ repo.  We currently have the following python bindings tests:
```
network_test.py
sparse_binary_matrix_test.py
sparse_matrix_test.py
sparse_link_test.py
cast_mode_test.py
pyregion_test.py
SDR_test.py
algorithms/cells4_test.py
check_test.py
array_algorithms_test.py
nupic_random_test.py
SDR_Metrics_test.py
temporal_memory_test.py
```
Notably missing: spatial-pooler tests, anomaly tests, connections tests

**Comment by dkeeney** [comment]
- One thing not yet on the list is conversion of the .py code to Python3.  This could be a big one.
I don't do much Python.  I think we should try to recruit some Python developers to do this task.

- I would also like to port some more of the encoders to C++.  The NetworkAPI does not have much support for plugging in flavors of encoder and I think there is some cool things we could do there.

- We talked about multi-threading 

- Another interesting idea is adapting the algorithms (SP and TM) for continuous persistence.  Rather than using serialization, identify the SP and TM data that identifies the minimum learned state and map that to a memory backed file.  Not everything like we do today with serialization, just the data that was previously learned so that if the program is ran again it could start up and continue learning based on what had been previously learned.

**Comment by breznak** [comment]
> the C++ SP? 

c++, I'm still thinking of this repo as cpp primarily 

> o Python3. This could be a big one. 

agree. But I think this will take a while. So maybe some later release?

> port some more of the encoders to C++ 

I may have some encoders for c++, will see in what state and if can be released. 

> adapting the algorithms (SP and TM) for continuous persistence

that is interesting? (probably OT, in another thread..but how would that work? I mean compared to RAM the speed would suffer)

**Comment by ctrl-z-9000-times** [comment]
>  I think we should try to recruit some Python developers to [port to python3].

I'm a python developer, I can take on this task.  I've done many py2-py3 conversions.

**Comment by ctrl-z-9000-times** [comment]
>  We talked about multi-threading

I don't think that this should be a requirement for releasing the software.  The software is usable without parallel execution.  It would be nice for performance though.

**Comment by dkeeney** [comment]
Ah 👍   That sounded like a volunteer for the Python work.

So we can sort of divide up things:
- @ctrl-z-9000-times  heads up python code
- @dkeeney  heads up NetworkAPI
- @breznak  heads up C++ algorithms

It would be nice to have a few more people involved too.

**Comment by breznak** [comment]
>  few more people involved too.

yea, badly. If we work on some nice feature, share it with others on the boards, I'm hoping we can pick people one by one on some tasks they are interested. (there's ton going on here just today!)

**Comment by Thanh-Binh** [comment]
Imho we SHOULD focus on c++ development as the name „nupic.cpp“ means, at least until the release 1.0. I see, you have very long todo-list with c++ algorithms.
If we want to support all, I think it is unnecessary double work so that we NEVER or very SLOW to reach the main goal „HTM should be better than any DL“.
What do you think?

**Comment by ctrl-z-9000-times** [comment]
Follow up

We should put together and agree on a response to inform all of the survey participants of our decisions.  Also, we never fully discussed/agreed on all of the things so now would be a good time for us to clarify our thoughts with each other.

This is high level / long term planning, so we should take our sweet time.  Only after following up with our customers should we put together a list of deliverables to work towards releasing.

**Comment by ctrl-z-9000-times** [comment]
Here is what I have so far:

### Follow Up

I want to thank everyone who voted and left comments.  We, the active developers of the community fork, have agreed on an agenda which we'd like to share with you.  These are tentative goals for the project:

 * An HTM which is accessible in C++ and Python
	- Will be implemented in C++, with bindings to python language.
	- Loose API compatibility with previous versions of Nupic.  We will change the API to improve the implementation and the usability of the algorithms. 
		+ Making & Integrating a class which handles Sparse Distributed Representations.  
		+ Rework how algorithms accept many parameters with structures?  I think this is too ambitious and breaks too many things.

 * The NetworkAPI
	- @dkeeney you should write this section since you're the one working on it, what I have here is merely suggestion...
	- Strict API compatibility with previous versions of Nupic?
	- What languages? C++ & Python

### Code Organization
We will endeavour to deliver a single code repository which contains the necessary tools in both C++ & Python.
 * However we won't copy the entire python repository into the C++ repository.  Instead we will copy files when they're needed and ready (meaning they pass a code review).
 * Not on this agenda: supporting duplicate algorithms in both C++ & Python.
 * Merging the unit-tests is our priority here, since they can be used to verify API compatibility

**Comment by dkeeney** [comment]
- The NetworkAPI
  - Strict API compatibility with previous versions of Nupic
  - To support:  Python, stand-alone C++, and C# interfaces.
  - access to more built-in algorithms and encoders  (C++ and do not require Python)
  - multi-threading; parallel execution of regions when possible.

**Comment by dkeeney** [comment]
The rest of it looks reasonable.  
I would like to include something about working with grid cells, displacements, and 1000 brain theory but I don't know what we would say at this point other than when algorithms become available we will incorporate them.

**Comment by ctrl-z-9000-times** [comment]
> grid cells, displacements, and 1000 brain theory

I have some thoughts about these things, but I don't want to go into the details in the **Follow Up** response.  I would rather add a few lines like:
 * Ongoing Research:
    - HTM theory is still being researched, and as new models are made and tested, we can polish them and bring them into this repository.

Here are the details of how I think we should deal with the subjects you bring up:

### Grid Cells

We know that the grid cells in the Entorhinal cortex should form hexagonal grids, and we can generate that pattern at will.  The process to generate the grids amounts to a couple of equations which act on the 2-D Euclidean Coordinates.  Numenta has an example of this somewhere?

Numenta does not have a biologically plausible model for grid cells.  Fear not, other researchers do.  I have experimentally reproduced the (Kropff & Treves, 2008) grid cell model and would like to contribute it to Nupic.  Without getting into the details: this model is highly compatible with current HTM theory.  It is an extension to the Spatial Pooler algorithm.  

### Displacement Cells

We can generate these given Numenta's descriptions.  Generating plausible displacement cells given the known real location of the robot is a simple encoding problem.

The true biological model of displacement cells is unknown.  I have a hypothesis about this; it's simple but contingent of having the (Kropff & Treves 2008) model.

### 1000 brains theory

This idea pertains to the large scale organization of the cortex.  The cortex looks pretty much the same everywhere so this isn't a theory about the small scale algorithms, this is a theory about how the areas are connected.  The NetworkAPI might be useful for implementing this theory, by helping to make & organize many brain regions.

**Comment by breznak** [comment]
>  We will change the API to improve the implementation and the usability of the algorithms.  

...while trying not to diverge too much from the original API. (and I'd skip the 2 bullets,details below) 

> ~~Strict~~Strong API compatibility with previous versions of Nupic

> I have some thoughts about these things, but I don't want to go into the details in the Follow Up response. 

I agree, I wouldn't go too technical/sciency in the follow-up (we don't want the discussion to turn to the research topic in that thread) 
I'm for not including the mentions of 1000brains, grid cells etc., You may put some links to our issues:
Interesting ongoing work: 
- parallelization
- column pooler, 
- improved spatial pooling
- vision experiments

We might put "Currently seeking help in the following areas": python merging, new c++ encoders, ...

**Comment by dkeeney** [comment]
> HTM theory is still being researched, and as new models are made and tested, we can polish them and bring them into this repository.

Yes, that sounds like a good way to say it.

>  Strict API compatibility with previous versions of Nupic

I think this is a reasonable thing to state in regards to the NetworkAPI.  Even if the algorithms change their calling interface, the NetworkAPI should not.

**Comment by ctrl-z-9000-times** [comment]
I incorporated your feedback.  Please give more feedback when able.  If you're satisfied give this post a thumbs up to approve.  After both of you have approved I will move it to the Numenta form.

### Follow Up (Revision 2)

I want to thank everyone who voted and left comments. We, the active developers of the community fork, have agreed on an agenda which we'd like to share with you. These are tentative goals for the next release of this project:
 * An HTM which is accessible in C++ and Python
      * Will be implemented in C++, with bindings to python language.
      * Loose API compatibility with previous versions of Nupic.  We will change the API to improve the implementation and the usability of the algorithms, while trying not to diverge too much from the original API.

  * The NetworkAPI
    * Strong API compatibility with previous versions of Nupic
    * To support:  Python, stand-alone C++, and C# interfaces.
    * Access to more built-in algorithms and encoders  (C++ and do not require Python)
    * Multi-threading; parallel execution of regions when possible.

 * Ongoing Research:      
      * HTM theory is still being researched, and as new models are made and tested, we can polish them and bring them into this repository.

### Code Organization

We will endeavour to deliver a single code repository which contains the necessary tools in both C++ & Python.
 * However we won't copy the entire python repository into the C++ repository.  Instead we will copy files when they're needed and ready (meaning they pass a code review).
 * Not on this agenda: supporting duplicate algorithms in both C++ & Python.
 * Merging the unit-tests is our priority here, since they can be used to verify API compatibility

**Comment by breznak** [comment]
> I incorporated your feedback. 

looks good to me, thanks! 
You might add a note that we'll be happy to welcome new testers and developers.

**Comment by breznak** [comment]
After publishing and waiting a couple days for some feedback, we may close this issue with a PR that updates with the "follow up" as our Agreed Strategy

**Comment by dkeeney** [comment]
looks good to me too.   Thanks for putting this together.

**Comment by breznak** [comment]
Closing this for now, may the force be with us!
